### PR TITLE
feat: support cancelling queued downloads if the UI is closed

### LINKED ISSUE #127: Cancel queued downloads if the UI is closed
### Feature proposal

After #117 is implemented, if for some reason the user that is waiting decided to close the browser or navigate away, it the UI is not present anymore, then the export should be canceled.

### Describe solution expectations

If there is an UI when `StreamResourceWriter::accept` is called, but the UI is no longer attached after the semaphore is acquired, then skip the document generation, with the expectation that there will be nobody to receive it. This approach fails when there is client-side navigation.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by javier-godoy**
#127 

## Summary by CodeRabbit

- **New Features**
	- Enhanced grid exporter to ensure UI attachment before performing export operations, improving overall stability and user experience.
	- Introduced configuration options to control behavior during UI changes, allowing for better error handling during downloads.
	- Added new methods to manage UI instances directly, facilitating improved interaction with the Vaadin UI context.
- **Bug Fixes**
	- Implemented stricter control over UI changes during export operations to prevent potential issues.
- **Tests**
	- Improved thread management and added new tests to verify behavior during concurrent export operations and UI changes.

### CODE REVIEW

**Comment by javier-godoy** [comment]
@coderabbitai review

File: src/test/java/com/flowingcode/vaadin/addons/gridexporter/ConfigurableConcurrentStreamResourceWriter.java [hunk 1]
```diff
@@ -32,4 +34,13 @@ public void setTimeout(long timeout) {
...
   }
 
+  @Override
+  public UI getUI() {
+    return ui;
+  }
+
+  public void setUi(UI ui) {
+    this.ui = ui;
+  }
```
[the remark this answers is not in the document]
Comment (javier-godoy) [reply src/test/java/com/flowingcode/vaadin/addons/gridexporter/ConfigurableConcurrentStreamResourceWriter.java:44]: @coderabbitai The local field is named `ui` (lowercase), so its setter is capitalized as `setUi`

**Review by mlopezFC — APPROVED** [review]
LGTM

**Comment by javier-godoy** [comment]
Resolved the conflicts. Also added a test for `setFailOnUiChange(true)` and made some minor improvements in the test setup.

**Review by paodb — APPROVED** [review]

**Pull request merged**